File: Common/src/main/java/cn/vove7/common/model/ResultBox.java
package cn.vove7.common.model;

/**
 * Created by Vove on 2018/7/5
 */
public class ResultBox<T> {
    private T mValue;

    public ResultBox() {
    }

    public ResultBox(T mValue) {
        this.mValue = mValue;
    }

    public T get() {
        return mValue;
    }

    public void set(T value) {
        this.mValue = value;
    }

    public void setAndNotify(T value) {
        mValue = value;
        synchronized (this) {
            notify();
        }
    }

    //等待结果
    public T blockedGet() {
        synchronized (this) {
            try {
                this.wait();
            } catch (InterruptedException e) {
                throw new RuntimeException(e);
            }
        }
        return mValue;
    }


}

File: common/src/main/java/cn/vove7/common/model/ResultBox.kt
package cn.vove7.common.model

import cn.vove7.common.utils.runInCatch
import cn.vove7.vtp.log.Vog
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit

/**
 * Created by Vove on 2018/7/5
 */
class ResultBox<T> {
    private var mValue: T? = null

    var lock = CountDownLatch(1)

    constructor()

    constructor(initValue: T) {
        this.mValue = initValue
    }

    fun setAndNotify(value: T) {
        Vog.d("setAndNotify ---> $value")
        mValue = value
        lock.countDown()
    }

    //等待结果
    @Throws(InterruptedException::class)
    fun blockedGet(safely: Boolean = true, millis: Long? = null): T? {
        if (lock.count <= 0) return mValue
        fun wait() {
            if (millis != null) lock.await(millis, TimeUnit.MILLISECONDS)
            else lock.await()
        }
        if (safely) runInCatch { wait() }
        else wait()
        return mValue
    }


}
